Provide `with_header` method on client

jongiddy opened the pull request:
This change allows arbitrary headers to be included in requests to ClickHouse.

It uses `HeaderName` and `HeaderValue` which is efficient for use with `hyper`. The types are provided by the `http` crate to allow support for other HTTP client crates to use these types without requiring the full `hyper` crate.

There are other ways to provide this capability, including unsealing the `HttpClient` trait to allow more implementations. However, this change provides a limited additional capability until a public interface for `HttpClient` is decided.

Fixes #98

slvrtrn commented:
In the hyper documentation, they use plain strings as headers names/values: https://docs.rs/hyper/latest/hyper/struct.Request.html#examples

> It uses HeaderName and HeaderValue which is efficient for use with hyper.

If there are optimizations made by HeaderName, and the custom headers will likely use some non-standard names (`X-My-Proxy-Auth` and similar), does it make sense to add `http` as a dependency, or maybe we can get away with just `&str`?

From https://docs.rs/http/latest/http/header/struct.HeaderName.html#representation:

> HeaderName represents standard header names using an enum, as such they will not require an allocation for storage. All custom header names are lower cased upon conversion to a HeaderName value. 

As I understand it, the custom header names are not represented as enums.

loyd commented on src/query.rs:
```diff
@@ -152,6 +152,10 @@ impl Query {
 
         let mut builder = Request::builder().method(method).uri(url.as_str());
 
+        for (name, value) in &self.client.headers {
```
What about inserts?

jongiddy replied:
There is a change in the `insert.rs` file to add the headers. Is this what you mean?

loyd commented:
LGTM, let's also pass to INSERTs

loyd commented on src/lib.rs:
````diff
@@ -170,6 +173,22 @@ impl Client {
+    /// ```
+    pub fn with_header(
+        mut self,
+        name: impl Into<HeaderName>,
````
I think we can accept `impl TryInto` to make it possible to pass strings. Downside here is that we cannot use `HeaderMap` and should use `HashMap<Result<_>, Result<_>>` instead =(

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/clickhouse.rs?pullRequest=108) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/clickhouse.rs?pullRequest=108) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/clickhouse.rs?pullRequest=108) it.</sub>

jongiddy commented:
I rebased the code on the latest version and rewrote it to use plain `String`s without requiring the `http` crate. The client headers now follow a similar structure to client options.

loyd commented:
Merged, thanks for your contribution!
